0005-ordered-quantification.md

```diff
@@ -0,0 +1,661 @@
+- Feature Name: ordered_quantification
+- Start Date: 2022-02-22 (twosday!)
+- RFC PR: [dafny-lang/rfcs#10](https://github.com/dafny-lang/rfcs/pull/10)
+- Dafny Issue: [dafny-lang/dafny#1753](https://github.com/dafny-lang/dafny/issues/1753)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes adding two closely-related new features to Dafny for working with collections and iteration:
+
+1. `foreach` loops over any enumerable domain, initially only supporting the builtin collection types but leaving the door open for user-defined collections.
```

Comment:
Have we considered introducing the concept of an iterator/enumerator type prior to introducing a `foreach`? The iterator/enumerator would be similar to that in Java/C#, and you could use it to do something similar to a `foreach`:

```dafny
method ForEach<T>(hasEnumerator: HasEnumerator<T>)
{
   var enum: BoundedEnumerator<T> := hasEnumerator.GetEnumerator();
   while(enum.HasNext()) 
      decreases enum.MaxLeft() {
      var iterationVariable := enum.MoveNext();
      // do something useful
   }
}

trait HasEnumerator<T> {
  method GetEnumerator() returns (result: BoundedEnumerator<T>)
}

trait BoundedEnumerator<T> {
   function MaxLeft(): int
   function method HasNext(): bool
   method MoveNext() returns (value: T)
      requires HasNext()
      ensures MaxLeft() < old(MaxLeft())
}
```

Building the `foreach` concept on top of such an iterator would mean we immediately enable user-defined collections to work with `foreach`. More importantly, starting out with simple fundamentals and building higher-level language constructs on top of those might mean we avoid having to backtrack in our implementation in the future.

Can you provide a sketch for how you plan to support `foreach` for user-defined collections?

---

Similarly, for supporting user-defined collections to be used in the quantified variable domain of a sequence comprehension, I expect we would need to define a trait 

```dafny
trait Foldable<T> {
  function method Fold<Result>(initial: Result, combine: (T, Result) -> Result): Result
}

function SequenceComprehension<T>(domain: Foldable<T>): seq<T> {
  domain.Fold([], (value: T, result: seq<T>) => [value] + result)
}
```

And for set comprehensions:

```dafny
function SetComprehension<T>(domain: Aggregatable<T>): set<T> 
   reads *
{
  domain.Select(v => {v}).Aggregate({}, (a,b) => a + b)
}

trait Aggregatable<T> {
  ghost var refs: set<T>;

  function method Select<U>(f: T -> U): Aggregatable<U>

  function method Aggregate(zero: T, aggregate: (T, T) -> T): T
   reads this
   requires forall x, y | x in refs && y in refs :: aggregate(x,y) == aggregate(y,x)
}

function FailedSeqComprehension<T>(domain: Aggregatable<T>): seq<T> 
   reads *
{
  domain.Select(v => [v]).Aggregate([], (a,b) => a + b)
}
```

Reply:
> Have we considered introducing the concept of an iterator/enumerator type prior to introducing a foreach?

Oh absolutely, and https://github.com/dafny-lang/libraries/pull/37 has a snapshot of that work to date. It looks similar to what you sketched out with an `Enumerator<T>` trait, and I'd expect an `Enumerable<T>` trait with a `GetEnumerator()` method. That would be the bare minimum to allow more values to be used as quantified variable domains.

There are a couple of factors that make that approach much harder to implement and less beneficial to users though, at least in the current state of Dafny:

1. Your `// do something useful` comment in `ForEach` skips over a big gap. :) Lambda expressions in Dafny can't have any side-effects, so you can't make a call like `ForEach(myCollection, x => { print x };` work. That means you have to declare a general-purpose "Action" trait and an explicit class for every use case, e.g. https://github.com/aws/aws-encryption-sdk-dafny/blob/c8be143088490301d486958ad126d53c46ec1311/src/AwsCryptographicMaterialProviders/Keyrings/AwsKms/AwsKmsKeyring.dfy#L599-L600. It's a massive pain and an order of magnitude harder to use than a built in `foreach` construct.
2. I found that attaching accurate enough termination metrics and invariants to the various `Enumerator<T>` implementations required creating a combinatorial explosion of concrete classes. I'm not against doing that in the standard library if necessary (and possibly leaning on code generation to make it easier/more maintainable), but I'd like to explore solving the more fundamental limitation first.
3. Consuming such enumerator types using a particular idiom of `while` loops (e.g. https://github.com/dafny-lang/libraries/pull/37/files#diff-1927bc3853704ac5f502a37509d998e3ac29f7fbe8c4e8c479810b6a02ca4c3bR15-R28) still takes a lot of boilerplate, and requires you to then prove your enumerator's `Repr` is disjoint from any other objects in your code.
4. Dafny also doesn't support having "default" implementations of methods in a trait that can be overridden in classes. That means we can't decide the `Enumerator<T>` interface should really have a `Contains(x)` method (for example) in the future without breaking existing code. Supporting default implementations was necessary in Java 8 so they could add the streams API to existing interfaces like `Collection` without breaking existing implementations.
5. Modelling the enumeration constraints as a trait is also limiting, since only reference types can currently implement them. The alternative is introducing yet more type characteristics similar to "failure-compatible" (and more complicated ones at that), and I'd rather avoid that and wait until https://github.com/dafny-lang/dafny/issues/1226 is addressed.

The appealing thing about the features in this RFC is that they fully abstract away from the details of enumerating at runtime. I absolutely plan for us to support user-defined collections in the future, but supporting ordered quantification over just the built-in collection types provides a ton of value while keeping the door open for the future. I've tried very hard to keep the semantics as general and flexible as possible, to reduce the risk you point out of finding that they interfere with how we'd like to define lower-level features that can implement them, but please let me know if you're concerned about anything specific.

Note that the aggregation/`Fold` functionality you also sketch out is out of scope for now. I suspect it will also work out better as a more built in concept as the "collect" future possibility sketches out, for a lot of the same reasons - aggregation is essentially the dual of enumeration!

Reply:
Okay, but how do you plan to support `foreach` for user-defined collections?

If that means introducing Enumerable<T> and rebuilding foreach on top of it, wouldn't that mean we lose some of the effort that we spend now on a first-class `foreach` implementation?

Reply:
The initial `foreach` implementation will translate pretty directly to the existing `BoundedPool` / `SinglePassCompiler.Create[Guarded]ForeachLoop/CompileGuardedLoops` implementations already used to compile features such as set comprehensions and assign-such-that statements, so the implementation cost beyond the language changes is actually quite cheap.

If we end up replacing that code with a more pure-Dafny implementation of builtin collections in the future, very little effort will be wasted. And it's likely sequence comprehensions will be useful in in the specification of such code!

Reply:
Okay, sounds good :-)